src/kernels/MIOpenSoftmaxAttn.cpp

```diff
@@ -436,3 +431,327 @@ extern "C" __global__ void ScaleReduce(const float* __restrict__ in,
         atomicMaxOfNonNegative(Amax, r_Amax);
     }
 }
+
+extern "C" __global__ void ScaleRowReduceWarp(const float* __restrict__ dO,
+                                              const float* __restrict__ O,
+                                              float* __restrict__ out,
+                                              const float* __restrict__ descale_dO,
+                                              const float* __restrict__ descale_O,
+                                              const float* __restrict__ dropout_P,
+                                              uint32_t d,
+                                              uint64_t nhs)
+{
+    constexpr uint32_t NumWarps = THREADS / warpSize;
+    const uint32_t lid          = threadIdx.x;
+    const uint32_t laneId       = lid % warpSize;
+    const uint32_t warpId       = lid / warpSize;
+    const float scaler          = (*descale_dO) * (*descale_O) * (1.0f - (*dropout_P));
+
+    for(uint64_t gid = blockIdx.x * NumWarps + warpId; gid < nhs; gid += gridDim.x * NumWarps)
+    {
+        float local_val = 0.0f;
+        if(laneId < d)
+        {
+            const float* dO_ptr = dO + gid * d + laneId;
+            const float* O_ptr  = O + gid * d + laneId;
+
+            local_val = (*dO_ptr) * (*O_ptr) * scaler;
+        }
+
+        local_val = reductionFullWarp<warpSize>(local_val, laneId, plus_op);
+
+        if(laneId == 0)
+        {
+            out[gid] = local_val;
+        }
+    }
+}
+
+extern "C" __global__ void ScaleRowReduceBlock(const float* __restrict__ dO,
+                                               const float* __restrict__ O,
+                                               float* __restrict__ out,
+                                               const float* __restrict__ descale_dO,
+                                               const float* __restrict__ descale_O,
+                                               const float* __restrict__ dropout_P,
+                                               uint32_t d,
+                                               uint64_t nhs)
+{
+    constexpr uint32_t NumWarps = THREADS / warpSize;
```

Comment:
[Q] Is it guaranteed that `THREADS % warpSize == 0`? If not, then we have to handle this case somehow (add 1 at least + possible more). Or this is guaranteed by the host code?

Reply:
It is "guaranteed" by the host code, but I'll add `static_assert` here.